### PR TITLE
Update contributing docs

### PR DESCRIPTION
**Pull request by rileytaylor**
# What's in this PR?

This includes a massive update to CONTRIBUTING.md that brings it in line with our API documentation. New users should be able to read through this and have a very good idea of how to get starting working on a task.

## References
Progress on: #423 
Fixes: #561

### CODE REVIEW

**Comment by JoshSmith** [comment]
@rileytaylor looks awesome. Will review thoroughly this afternoon.

**Comment by rileytaylor** [comment]
I believe this covers everything.

The only thing we don't really talk about is branch formatting ( we had discussed trying to shift toward issue#-add|fixes|removes-thing, which is what we've all mostly been using ) I'm not sure where a good place to put it would be. My only thinking within this document would be to add a "Before you make any changes to the codebase" section. We could potentially make a new file on "How we use git" or make the SQUASHING.md more about git use in general than just rebasing... but that may make it a bit unwieldy.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -1,38 +1,148 @@
...
+
+### Improving Ember docs
+
+You can see how much code is documented on [Inch CI](http://inch-ci.org/github/code-corps/code-corps-ember).
+
+We use YUIDoc to document our code.
+
+You can learn about how to write documentation with YUIDoc from these places:
+
+- Our [Styleguide](../docs/STYLEGUIDE.md#Documentation)
```
Comment (JoshSmith) [inline .github/CONTRIBUTING.md]: Prefer "style guide"

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -1,38 +1,148 @@
...
+- Open an issue first. It's better if we discuss your proposed changes.
+- We try to keep the main `README.md` lightweight and use it as a jumping point to other docs.
+- Most other docs can be placed in `/docs`.
+- Try to make it easy for people to jump around in your doc.
+
+[Done with your changes?](#i-finished-my-changes)
+
+## How to fix a bug
+
+If you're fixing a bug that's already been added to the issues, ask yourself whether the bug description is clear? Do you know what circumstances led to the bug? Does it seem easy to reproduce?
```
Comment (JoshSmith) [inline .github/CONTRIBUTING.md]: First line should be question. May be wrong in other docs, too.

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -1,38 +1,148 @@
...
+
+If you've spotted a bug yourself, open an issue and try to answer those questions.
+
+Then start writing some code:
+
+1. Make the tests fail.
+
+   Identify what's happening in the bug with a test. This way the bug is reproducible for everyone else in the project, and we won't regress into making the bug ever again (hopefully!).
```
Comment (JoshSmith) [inline .github/CONTRIBUTING.md]: Remove these tabs.

File: .github/CONTRIBUTING.md (hunk 2 above)
Comment (rileytaylor) [inline .github/CONTRIBUTING.md]: Should? or shouldn't? I think it should be:
>...ask yourself whether the bug description is clear. Do you know what circumstances led to the bug?...

If we want it to be a question then it'll need a rewrite.

**Comment by rileytaylor** [comment]
Just one question on those; updated the other two things. I'm also still debating where to put branch naming and recommending `hub`.

**Comment by JoshSmith** [comment]
@rileytaylor I think just put it anywhere for now. Better to have it than not; if people don't use it / understand it, we'll hear.

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -146,3 +150,59 @@ The CircleCI builds also rely on some environment variables for reporting, deplo
...
+
+When making code changes with Git, it is best to develop a workflow. Every 'feature', which can be anything from a 1-line change to an entire component, should be done on a seperate feature branch.
+
+Let's say you are working on issue 6000, which requests that the `needs_coffee` property of `programmer` be changed to a constant that always returns `true`. You'll make the following steps:
+
+1. Check out the code, or fetch any updates from github.
+
+2. Assign yourself the issue (if you aren't a contributor yet, comment that you are working on it)
+
+3. Create a new feature branch:
```
Comment (JoshSmith) [inline .github/CONTRIBUTING.md]: Should be able to remove the separate line breaks without issue here I think.

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -146,3 +150,59 @@ The CircleCI builds also rely on some environment variables for reporting, deplo
...
+
+[Install `posh-git`](https://github.com/dahlbyk/posh-git) which, like bash git completion, provides tab completion and helpful repository info in your prompt.
+
+You'll probably want to keep some sort of Bash emulator on hand. Git comes with it's own bash shell, but its not very good. Fortunately there are a few other great options:
+
+- [ConEmu](http://conemu.github.io/) is a fantastic console emulator that comes with two bash options.
+- [Cmder](http://cmder.net/) is ConEmu on steroids with Monokai console coloring and git bash completion, Posh-git, and oh-my-posh already installed
+- [Windows Subsystem for Linux](https://blogs.msdn.microsoft.com/wsl/2016/04/22/windows-subsystem-for-linux-overview/) More complex to use, but it gives you a full bash shell in an Ubuntu enviornment.
+
+
```
Comment (JoshSmith) [inline .github/CONTRIBUTING.md]: Extra lines here.

**Comment by JoshSmith** [comment]
Good to go after those changes.

**Pull request merged**